server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/ContinuousComputation.java

```diff
@@ -46,6 +49,20 @@ public void onNewInput(T input) {
         }
     }
 
+    public void onNewInput(T input, Runnable listener) {
+        assert input != null;
+
+        boolean executeProcessor;
+        synchronized (this) {
+            executeProcessor = enqueuedInput.getAndSet(Objects.requireNonNull(input)) == null;
+            listeners.get().add(listener);
```

Comment:
Since this is synchronized, atomics are no longer needed, however I am checking if it is possible to avoid sync here or implement it other way